driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -101,135 +114,98 @@ public InternalConnection get() {
+            connectionPoolListener.connectionCheckOutFailed(
+                    new ConnectionCheckOutFailedEvent(serverId, reason == null ? Reason.UNKNOWN : reason));
         }
     }
 
     private void openAsync(final PooledConnection pooledConnection,
-                           final SingleResultCallback<InternalConnection> callback) {
+                           final SingleResultCallback<InternalConnection> callback,
+                           final Timeout timeout) {
         if (pooledConnection.opened()) {
             if (LOGGER.isTraceEnabled()) {
                 LOGGER.trace(format("Pooled connection %s to server %s is already open",
-                                           pooledConnection.getDescription().getConnectionId(), serverId));
+                        getId(pooledConnection), serverId));
             }
-            connectionPoolListener.connectionCheckedOut(
-                    new ConnectionCheckedOutEvent(pooledConnection.getDescription().getConnectionId()));
             callback.onResult(pooledConnection, null);
         } else {
             if (LOGGER.isTraceEnabled()) {
                 LOGGER.trace(format("Pooled connection %s to server %s is not yet open",
-                                           pooledConnection.getDescription().getConnectionId(), serverId));
+                        getId(pooledConnection), serverId));
             }
-            pooledConnection.openAsync(new SingleResultCallback<Void>() {
-                @Override
-                public void onResult(final Void result, final Throwable t) {
-                    if (t != null) {
-                        if (LOGGER.isTraceEnabled()) {
-                            LOGGER.trace(format("Pooled connection %s to server %s failed to open",
-                                                       pooledConnection.getDescription().getConnectionId(), serverId));
-                        }
-                        connectionPoolListener.connectionCheckOutFailed(new ConnectionCheckOutFailedEvent(serverId,
-                                Reason.CONNECTION_ERROR));
-                        callback.onResult(null, t);
-                        pool.release(pooledConnection.wrapped, true);
-                    } else {
-                        if (LOGGER.isTraceEnabled()) {
-                            LOGGER.trace(format("Pooled connection %s to server %s is now open",
-                                                       pooledConnection.getDescription().getConnectionId(), serverId));
-                        }
-                        connectionPoolListener.connectionCheckedOut(
-                                new ConnectionCheckedOutEvent(pooledConnection.getDescription().getConnectionId()));
-                        callback.onResult(pooledConnection, null);
-                    }
-                }
-            });
+            getAsyncGetter().execute(() -> openConcurrencyLimiter.openOrGetAvailableAndSilentlyCloseSpecified(
```

Comment:
[Thoughts about the current (before the PR) design of asynchronous pool operations]

While it is not the most efficient approach to implement asynchrony here by running a blocking task in a different thread created specifically for that purpose, this approach has already been used by the pool, and I decided to reuse it for the new functionality.

An alternative approach for asynchronous `openOrGetAvailableAndSilentlyCloseSpecified`:
- if a pool can provide an open connection without waiting, then complete the callback synchronously;
- otherwise store the callback in a queue and complete when either a permit is released or an available open connection appears.

The same approach could have been implemented for checking out connections from the pool. But it would cause threads that release connections to complete checkout callbacks, which would make release latencies unpredictable<sup>1</sup> in addition to checkout latencies being unpredictable<sup>1</sup> (this is currently true).

---
<sup>1</sup> This is determined by the callbacks pool receives.